### ISSUE TITLE
Download if url with spaces

### ISSUE THREAD

**Issue by DhruvaSambrani** [JuliaLang/julia#37727]
# Info
Any `download("url with spaces")` fails with `The requested URL returned error: 400 Bad Request`

# versioninfo
```julia 
Julia Version 1.5.1
Commit 697e782ab8 (2020-08-25 20:08 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)
```

# MWE
```julia
julia> download("http://www.glozman.com/TextPages/Harry Potter 2 - Chamber of Secrets.txt")
┌ Error: Download failed: curl: (22) The requested URL returned error: 400 Bad Request
└ @ Base download.jl:43
ERROR: failed process: Process(`'C:\WINDOWS\System32\curl.exe' -s -S -g -L -f -o 'C:\Users\dhruv\AppData\Local\Temp\jl_ncRp4v9UgV' 'http://www.glozman.com/TextPages/Harry Potter 2 - Chamber of Secrets.txt'`, ProcessExited(22)) [22]
```

# Possible fix
URL encode, or simply call curl with another set of `"` marks

# Workaround
Replace all spaces with %20 or use [`HTTP.URIs.escapeuri`](https://juliaweb.github.io/HTTP.jl/dev/public_interface/#HTTP.URIs.escapeuri)

**Comment by KristofferC** [comment]
Could you change the example to not link to copyrighted material.

**Comment by DhruvaSambrani** [comment]
So I searched for a url with this issue, but could only find `https://www.google.com/search?q=hi what is up`. With the spaces it throws a `400 Bad Request` error and with %20 escaping, it throws a `403 Forbidden` because Google doesn't allow automated queries. However, it still illustrates the point.

**Comment by DhruvaSambrani** [comment]
Also sorry for the copyrighted link 😅

**Comment by DhruvaSambrani** [comment]
Would you prefer that I open an Issue in JuliaLang/Downloads? I assumed that the code would be in stdlib, but later realised its a different repo

**Comment by KristofferC** [comment]
I'm not sure where it's best. cc @StefanKarpinski

**Comment by StefanKarpinski** [comment]
URLs cannot have spaces in them (see https://en.wikipedia.org/wiki/URL). I'm fairly reluctant to automatically URL encode a URL that has spaces in it. Do we URL-encode the whole thing? That would turn
```
https://www.ecosia.org/search?q=c dij dqwd
```
into
```
https%3a%2f%2fwww.ecosia.org%2fsearch%3fq%3dc%20dij%20dqwd
```
which is not what you want either. So you want to URL encode some characters but not others? Which characters should be encoded? Which should be left alone? Is there some standard here?

**Comment by DhruvaSambrani** [comment]
Javascript's [`encodeURI`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI) and the associated [RFC2396](https://www.ietf.org/rfc/rfc2396.txt) may be the standard that is followed.

Or we could take a safe approach and just replace a space with a `+`

**Comment by StefanKarpinski** [comment]
I don't want to implement something ad hoc here. Implementing `encodeURI` would be fine though. I already have an implementation of full escaping here:

https://github.com/JuliaLang/Downloads.jl/blob/e367c6d13568a2cfca57286939e312d59aeb559d/test/setup.jl#L61-L83

I think the only difference is which characters get escaped.

**Comment by DhruvaSambrani** [comment]
This is mozilla's MDNS page on it
https://developer.mozilla.org/en-US/docs/Glossary/percent-encoding

**Comment by DhruvaSambrani** [comment]
No, ok, I see the issue now. All percent encoding is supposed to be done to the data of the url, not the url itself. Should we then parse the url for the data and then urlencode it, and then put it back together to send to Curl? Or is this a fringe case which is not worth the time of basically rewriting Python urllib?
I suppose this decision can be taken by mods as they see fit.

**Comment by StefanKarpinski** [comment]
`encodeURI` does not encode any of the characters that are significant to URLs. Something simple like that seems possible. Otherwise how do you know if a `/` is part of the URL or data? There's also [punycode](https://en.wikipedia.org/wiki/Punycode) for the hostname part, which is how Unicode domain names are generally handled. If we URL encode non-ASCII characters in the whole URL, we'll end up mangling domains in a way that's not compatible with that and will try to lookup the wrong domain name — `%` isn't even valid in a domain. So it seems like the "fully correct" version of this would be to parse the domain out of the URL, punycode that, and then percent-encode the whole resulting URL, skipping URL-significant characters.

But that's pretty darned complicated and if we implement this and get it wrong, then we can't change it again because that would be a breaking change. So if we're going to do this, I would be much happier with an explicit standard or at least some other implementations are references that what we're doing is correct.

**Comment by StefanKarpinski** [comment]
The wikipedia page on URLs actually has a section on this:

https://en.wikipedia.org/wiki/URL#Internationalized_URL

This basically matches what I described, but it would be good to have a standard or reference implementation.

**Comment by StefanKarpinski** [comment]
This seems to be the relevant standard: https://tools.ietf.org/html/rfc3987; unfortunately it's quite complex.
